src/pds/aipgen/registry.py

```diff
@@ -0,0 +1,517 @@
+    characters off the beginnings of PDS paths, and using info in the ``bac``. Return a triple of
+    the MD5 of the created manifest, its size in bytes, and a count of its entries.
+    '''
+    _logger.debug('⚙️ Writing AIP transfer manifest to %s', fn)
+    hashish, size, count = hashlib.new('md5'), 0, 0
+    with open(fn, 'wb') as o:
+        for lidvid, files in bac.items():
+            for f in files:
+                entry = f'{lidvid:255}/{f.url[prefixLen:]:255}\r\n'.encode('utf-8')
+                o.write(entry)
+                hashish.update(entry)
+                size += len(entry)
+                count += 1
+                if count % _progressLogging == 0:
+                    _logger.debug('⏲ Wrote %d entries into the transfer manifest %s', count, fn)
+    _logger.info('📄 Wrote AIP transfer manifest: %s', fn)
+    return hashish.hexdigest(), size, count
+
+
+def _writeAIP(bundleLIDVID: str, prefixLen: int, bac: dict, ts: datetime) -> str:
+    '''Create the PDS Archive Information Package for the given ``bundleLIDVID``, stripping
+    ``prefixLen`` characters off file paths and using information in the ``bac``.  The ``ts``
+    timestamp tells what metadata to put in the PDS label and the date for generated PDS
+    filenames. Return a stringified version of the MD5 hash of the *checksum manifest* of the AIP.
+    '''
+    _logger.debug('⚙️ Creating AIP for %s', bundleLIDVID)
+    cmFN = _makeFilename(bundleLIDVID, ts, 'checksum_manifest', PDS_TABLE_FILENAME_EXTENSION)
+    tmFN = _makeFilename(bundleLIDVID, ts, 'transfer_manifest', PDS_TABLE_FILENAME_EXTENSION)
+    cmMD5, cmSize, cmNum = _writeChecksumManifest(cmFN, prefixLen, bac)
+    tmMD5, tmSize, tmNum = _writeTransferManifest(tmFN, prefixLen, bac)
+    lid, vid = _deURNLIDVID(bundleLIDVID)
+    labelFN = _makeFilename(bundleLIDVID, ts, 'aip', PDS_LABEL_FILENAME_EXTENSION)
+    writeAIPLabel(labelFN, f'{lid}_v{vid}', lid, vid, cmFN, cmMD5, cmSize, cmNum, tmFN, tmMD5, tmSize, tmNum, ts)
+    _logger.info('📄 Wrote label for them both: %s', labelFN)
+    return cmMD5
+
+
+def _writeSIP(bundleLIDVID: str, bac: dict, title: str, site: str, ts: datetime, cmMD5: str):
+    '''Write a Submission Information Package based on the ``bac`` to the current directory
+    generating PDS filenames and other label metadata from the timestamp ``ts`` and ``bundleLIDVID``.
+    The ``cmMD5`` is the MD5 digest of the PDS Archive Information Package's transfer manifest and
+    also goes into the PDS label. The PDS ``site`` is a string like ``PDS_ATM`` indicating the
+    PDS site. You'd think we could get that from the PDS API but 🤷‍♀️.
+    '''
+    _logger.debug('⚙️ Creating SIP for %s (title %s) for site %s', bundleLIDVID, title, site)
+    sipFN = _makeFilename(bundleLIDVID, ts, 'sip', PDS_TABLE_FILENAME_EXTENSION)
+    hashish, size, count = hashlib.new('md5'), 0, 0
+    with open(sipFN, 'wb') as o:
```

Comment:
Third time seeing a variation of this hash/file generation loop, if the format string assigned to `entry` became a parameter itself, they could probably all be combined into a single helper function. Just something to consider.

Reply:
There are 3 such loops, 2 of which use the same criteria (iteration over lidvid → set of files) but one which is just different (set of files). If the 2 become 3, I'll refactor, but for now I'm okay to live with it as is.